test/unit/org/apache/cassandra/io/sstable/SSTableUniqueIdentifierTest.java

```diff
@@ -0,0 +1,92 @@
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.quicktheories.QuickTheory.qt;
+import static org.quicktheories.generators.SourceDSL.longs;
+
+public class SSTableUniqueIdentifierTest
+{
+    @Test
+    public void testSequenceBasedIdProperties()
+    {
+        testSSTableUniqueIdentifierProperties(SequenceBasedSSTableUniqueIdentifier.Builder.instance);
+    }
+
+    @Test
+    public void testUUIDBasedIdProperties()
+    {
+        testSSTableUniqueIdentifierProperties(UUIDBasedSSTableUniqueIdentifier.Builder.instance);
+    }
+
+    private void testSSTableUniqueIdentifierProperties(SSTableUniqueIdentifier.Builder<?> builder)
+    {
+        List<SSTableUniqueIdentifier> ids = Stream.generate(builder.generator(Stream.empty()))
+                                                  .limit(100).collect(Collectors.toList());
+        assertThat(ids).isSorted();
+        assertThat(Sets.newHashSet(ids)).hasSameSizeAs(ids);
+
+        List<ByteBuffer> serIds = ids.stream().map(SSTableUniqueIdentifier::asBytes).collect(Collectors.toList());
+        assertThat(serIds).isSortedAccordingTo((o1, o2) -> UnsignedBytes.lexicographicalComparator().compare(o1.array(), o2.array()));
+
+        List<SSTableUniqueIdentifier> deserIds = serIds.stream().map(builder::fromBytes).collect(Collectors.toList());
+        assertThat(deserIds).containsExactlyElementsOf(ids);
+
+        List<String> stringifiedIds = ids.stream().map(SSTableUniqueIdentifier::asString).collect(Collectors.toList());
+        if (!(builder instanceof SequenceBasedSSTableUniqueIdentifier.Builder))
+        {
+            // the legacy string representation is not sortable
+            assertThat(stringifiedIds).isSorted();
+        }
+
+        List<SSTableUniqueIdentifier> destringifiedIds = stringifiedIds.stream().map(builder::fromString).collect(Collectors.toList());
+        assertThat(destringifiedIds).containsExactlyElementsOf(ids);
+    }
+
+    @Test
+    public void testUUID()
+    {
+        qt().forAll(longs().all(), longs().all()).checkAssert((msb, lsb) -> {
+            msb = (msb & ~0xf000) | 0x1000; // v1
+            UUID uuid = new UUID(msb, lsb);
+            assertThat(uuid.version()).isEqualTo(1);
+            UUIDBasedSSTableUniqueIdentifier id = new UUIDBasedSSTableUniqueIdentifier(uuid);
+            ByteBuffer buf = id.asBytes();
+            UUIDBasedSSTableUniqueIdentifier fromBytes = UUIDBasedSSTableUniqueIdentifier.Builder.instance.fromBytes(buf);
+            assertThat(fromBytes).isEqualTo(id);
+            String s = id.asString();
+            assertThat(s).hasSize(UUIDBasedSSTableUniqueIdentifier.STRING_LEN);
+            assertThat(s).matches(Pattern.compile("[0-9a-z]{4}_[0-9a-z]{4}_[0-9a-z]{18}"));
+            UUIDBasedSSTableUniqueIdentifier fromString = UUIDBasedSSTableUniqueIdentifier.Builder.instance.fromString(s);
+            assertThat(fromString).isEqualTo(id);
+        });
```

Comment:
The test is a bit hard to read. I would split it in to sub-parts like `testBytesSerialization` and `testStringSerialization`.
It would also be good to check that the underlying UUID is still equals to the original one.